File: 3383-taking-maximum-energy-from-the-mystic-dungeon/taking-maximum-energy-from-the-mystic-dungeon.py
class Solution:
    def helper(self, i, energy, k):
        if i < 0: return 0
        prev = self.helper(i-k, energy, k)
        return energy[i] + prev if prev > 0 else energy[i]
        

    def maximumEnergy(self, energy: List[int], k: int) -> int:
        n = len(energy)
        res = float("-inf")
        for i in range(n-1,n-1-k,-1):
            if i >= 0:
                res = max(res, self.helper(i,energy, k))
            else:
                break
        return res

File: 3383-taking-maximum-energy-from-the-mystic-dungeon/taking-maximum-energy-from-the-mystic-dungeon.cpp
class Solution {
public:
    int maximumEnergy(vector<int>& energy, int k) {
        int n = energy.size();
        vector<int> dp(energy.size(),-1);
        int ans=INT_MIN;
        for(int i=energy.size()-1;i>=0;i--)
        {
            if(i+k<energy.size())
            {
                dp[i]=energy[i]+dp[i+k];
            }
            else
            {
                dp[i]=energy[i];
            }
            ans=max(ans,dp[i]);
        }
        return ans;
    }
};